src/runtime-rs/src/shim/src/lib.rs

```diff
@@ -0,0 +1,311 @@
+            || self.address.is_empty()
+            || self.publish_binary.is_empty()
+        {
+            return Err(Error::InvalidArgument);
+        }
+        if is_delete && self.bundle.is_empty() {
+            return Err(Error::InvalidArgument);
+        }
+
+        if Self::is_component_dangerous(&self.id) || Self::is_component_dangerous(&self.namespace) {
+            return Err(Error::InvalidArgument);
+        }
+
+        // Ensure `address` is a valid path.
+        let path = PathBuf::from(self.address.clone())
+            .canonicalize()
+            .map_err(|_| Error::InvalidArgument)?;
+        let md = path.metadata().map_err(|_| Error::InvalidArgument)?;
+        if !md.file_type().is_socket() {
+            return Err(Error::InvalidArgument);
+        }
+        self.address = path
+            .to_str()
+            .map(|v| v.to_owned())
+            .ok_or(Error::InvalidArgument)?;
+
+        // Ensure `bundle` is a valid path.
+        if !self.bundle.is_empty() {
+            let path = PathBuf::from(self.bundle.clone())
+                .canonicalize()
+                .map_err(|_| Error::InvalidArgument)?;
+            let md = path.metadata().map_err(|_| Error::InvalidArgument)?;
+            if !md.is_dir() {
+                return Err(Error::InvalidArgument);
+            }
+            self.bundle = path
+                .to_str()
+                .map(|v| v.to_owned())
+                .ok_or(Error::InvalidArgument)?;
+        }
+
+        Ok(())
+    }
+
+    fn is_component_dangerous(comp: &str) -> bool {
+        comp.contains(' ') || comp.contains('/') || comp.contains('\t')
+    }
+}
+
+/// Command executor for shim.
+pub struct ShimExecutor {
+    args: ShimArgs,
+}
+
+impl ShimExecutor {
+    /// Create a new instance of [`Shim`].
+    pub fn new(args: ShimArgs) -> Self {
+        ShimExecutor { args }
+    }
+
+    fn load_image_spec(&self) -> Result<oci_spec::runtime::Spec> {
+        let bundle_path = self.get_bundle_path()?;
+        let spec_file = bundle_path.join("config.json");
+
+        oci_spec::runtime::Spec::load(spec_file).map_err(Error::OciSpec)
+    }
+
+    fn write_address(&self, address: &str) -> Result<()> {
+        let dir = self.get_bundle_path()?;
+        let file_path = &dir.join("address");
+        std::fs::write(file_path, address.as_bytes())
+            .map_err(|e| Error::WriteFile(e, file_path.clone()))
+    }
+
+    fn write_pid_file(&self, pid: &str) -> Result<()> {
+        let dir = self.get_bundle_path()?;
+        let file_path = &dir.join("shim.pid");
+        std::fs::write(file_path, pid.as_bytes())
+            .map_err(|e| Error::WriteFile(e, file_path.clone()))
+    }
+
+    fn read_pid_file(&self, bundle_path: &Path) -> Result<String> {
+        let file_path = bundle_path.join("shim.pid");
+        // Limit the size of content to read in.
+        let size = file_path.metadata().unwrap().len();
+        if size > 32 {
```

Comment:
Why?

Reply:
Here we just check the file length to "weakly" make sure the pid file does include invalid content. Change 32 to 20, which is the maximum number of decimal digits for u64